Ble extended advertising fixes

pan- opened the pull request:
### Description

Second round of fixes discovered with the OOB and our testing.

It fixes: https://github.com/ARMmbed/mbed-os-example-ble/issues/198 and many other bug discovered during our internal testing on the Cordio LL.

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

pan- committed:
BLE: Revert change in stack setup initialisation.

While good this change breaks some old application that were adding GATT services before the initialisation of BLE.

This patch revert temporarily this change for now

pan- committed:
BLE: Fix conversion of advertising type

Previously, we were passing the event_properties as defined by the Bluetooth spec which is not what DmAdvConfig expect as the advertising type passed to DmAdvConfig is Cordio tailored (and incomplete).

pan- committed:
BLE: Set advertising random address when appropriate.

paul-szczepanek-arm committed:
fix return value

paul-szczepanek-arm committed:
check return value of DmSyncStart

paul-szczepanek-arm committed:
fix swapped errors

paul-szczepanek-arm committed:
always set filter policy

paul-szczepanek-arm committed:
copy periodic payload

pan- committed:
BLE: Allow null value for periodic interval in advertising report event.

pan- committed:
BLE: Fix address type allowed to create a periodic sync.

pan- committed:
BLE: Fix recursion in ble::advertising_data_status_t raw constructor.

pan- commented:
@ARMmbed/mbed-os-maintainers Fix for https://github.com/ARMmbed/mbed-os-example-ble/issues/200 will be included in that branch.

cmonr commented on features/FEATURE_BLE/ble/gap/Events.h:
```diff
@@ -63,7 +63,7 @@ struct AdvertisingReportEvent {
         advertising_power_t txPower,
         rssi_t rssi,
-        periodic_interval_t periodicInterval,
+        uint16_t periodicInterval,
```
Odd. Why not replace/update the constructor instead?

pan- replied:
As you can see, that constructor is not part of the documentation because it is not meant to be used by user code. We cannot use `friend` in this case because `friend` cannot be inherited and we don't know the vendor file that will use it. There is no reason for us to keep non usable constructor  if it is not part of the public API. 

To give a bit more context, the periodic interval returned in an advertising event can be equal to 0 if there is no periodic interval or a value comprised between [0x0006 : 0xFFFF]. The type `periodic_interval_t` is not able to represent the absence of value as it normalizes values passed in input.

cmonr commented on features/FEATURE_BLE/targets/TARGET_CORDIO/source/CordioPalGap.cpp:
```diff
@@ -911,12 +984,11 @@ ble_error_t Gap::extended_scan_enable(
             DM_DISC_MODE_NONE,
             extended_scan_type,
-            filter_duplicates.value(), // TODO: cordio API incomplete ???
+            filter_duplicates.value(),
```
Any more TODOs to update? 😜

pan- replied:
I'm not sure why I removed it, the comment still stands 😆 . These TODO's and FIXME are important in our implementations and internal files when we work with APIs that do not implement all Bluetooth requirements or when we deliver subsets of BT features.

cmonr reviewed (commented):
LGTM, just a couple of questions.

pan- committed:
BLE: Revert changes introduced by debugging.

donatieng commented on features/FEATURE_BLE/targets/TARGET_CORDIO/stack/cordio_stack/ble-host/sources/stack/dm/dm_adv_ae.c:
```diff
@@ -1634,13 +1634,13 @@ void DmPerAdvSetData(uint8_t advHandle, uint8_t op, uint8_t len, uint8_t *pData)
     pMsg->op = op;
     pMsg->len = len;
-    pMsg->pData = pData;
+    memcpy(pMsg->pData, pData, len);
```
Let's make sure we upstream this

donatieng reviewed (approved):
Looks good to me!

adbridge commented:
@cmonr are you happy with Vincent's responses ? We should get this into ci anyway I think

pan- committed:
BLE: Add an option to inject the random static address during the reset sequence.

pan- committed:
BLE: Inject random static address during reset sequence.

pan- committed:
BLE: Set the number of supported phy by the host to 3.

pan- committed:
BLE: Set the number of the advertising sets supported by the host to 3.

pan- committed:
BLE: Fix the number of advertising sets supported

The number of advertising sets supported is the minimum of advertising sets supported beween the host and the controller.

adbridge commented:
CI started

adbridge commented:
CI stopped again awaiting Vincent's last fixes

paul-szczepanek-arm committed:
baseband clock rate increased to 1 000 000

paul-szczepanek-arm committed:
added arm version of libs

paul-szczepanek-arm committed:
iar libs for cordio

adbridge commented:
@donatieng could you please re-review

pan- commented:
@adbridge There's an issue with the libs generated, they will be re generated shortly.

cmonr commented:
@pan- Ping @ARMmbed/mbed-os-maintainers when this is ready for CI again.

Commit:
Remove duplicate symbols in libcordio_stack.a

Commit:
Remove use of GPIOs (LEDs and diag pins) from Cordio LL for Nordic

Commit:
ARMCC only likes armar

Commit:
Replace ARMCC libs

paul-szczepanek-arm committed:
working libs for GCC and IAR

0xc0170 commented:
CI started

0xc0170 commented:
CI aborted, will wait for @ARMmbed/mbed-os-pan  to confirm this is ready for CI (noticed the last commit after CI started)

paul-szczepanek-arm commented:
sorry, still not ready we're have to figure out the ARM compiler problem first

paul-szczepanek-arm commented:
ARM cc fix incoming within the hour

paul-szczepanek-arm committed:
working ARM lib

paul-szczepanek-arm committed:
fixed ARM compilation problem caused by noreturn

paul-szczepanek-arm commented:
wait

paul-szczepanek-arm committed:
remove noreturn error function to avoid compilation issues

paul-szczepanek-arm commented:
good to go

0xc0170 commented:
CI started

kjbracey commented:
What actually was the problem here? Compiler deducing that `ble_cordio_get_hci_driver` didn't return by looking at the weak version?

Was initially puzzled because assert failures are also noreturn, but then this assert is true so doesn't do anything.

I would have thought the code should be issuing an error in the calling code if `ble_cordio_get_hci_driver` returns NULL (whether it is because it fell through to the weak code, or because a real implementation had some other init problem). That's the pattern we've used in all the `get_default_instance` patterns - any base weak version is just a plain "return NULL".

pan- commented:
> What actually was the problem here? Compiler deducing that ble_cordio_get_hci_driver didn't return by looking at the weak version?

Yes, a straight call to error seems to taint the weak function as a no return function and that attributes propagates seems to be propagated by the linker to the non weak version. More investigation would be required to find out the issue details but we didn't had the time to conduct that full investigation as we were urged to find a solution. 

As you pointed out, the assert statement in the weak function is not correct as it always return true. The pattern you described seems suited for our use and may be a long term fix (not sure of the behaviour if the weak factory and the non weak accessor are in the same source file ...).

kjbracey commented:
Your pattern is almost totally equivalent to what we've done in networking and other areas, so it would be nice to align it.

After a couple of slightly varying iterations (1 layer/2 layer, pointer/reference), the final refined form I would suggest for you is

```
MBED_WEAK static BLEInstanceBase *BLEInstanceBase::get_default_instance()
{
     return get_target_default_instance();
}

MBED_WEAK static BLEInstanceBase *BLEInstanceBase::get_target_default_instance()
{
     return NULL;
}
````
Code gated on TARGET_XXX (only) gets to override the `get_target_default_instance()`.

Application or plug-in module code has the option of overriding `get_default_instance()`, and the extra layer ensures that whatever the target is doing  doesn't interfere with that. A library acting as a module plugin driver like ESP8266 would provide a JSON option `xxx.provide-default` to do that override.
